MaxText/inference_utils.py

```diff
@@ -172,3 +174,77 @@ def sample_topk_logits(logits, topk, temperature, rng):
   topk_token = jnp.expand_dims(jax.random.categorical(rng, topk_logits / temperature).astype(jnp.int32), axis=-1)
   sampled_tokens = jnp.squeeze(jnp.take_along_axis(topk_idxs, topk_token, axis=-1), axis=-1).astype(jnp.int32)
   return sampled_tokens
+
+
+def sample_topk_topp_weighted(logits, topk, nucleus_topp, temperature, rng):
+  """Applies top-k, top-p, and temperature sampling to logits.
+
+  This function combines three common sampling techniques to control the
+  randomness and diversity of the generated text. The operations are applied
+  sequentially:
+
+  1.  **Top-k filtering**: The vocabulary is restricted to the `topk` most
+      likely tokens.
+  2.  **Top-p (nucleus) filtering**: From the `topk` tokens, the smallest
+      set of tokens whose cumulative probability exceeds `nucleus_topp` is
+      selected.
+  3.  **Temperature scaling**: The logits of the filtered tokens are scaled
+      by the `temperature`. Higher temperatures result in a flatter
+      distribution (more randomness), while lower temperatures make the
+      distribution sharper (less randomness).
+  4.  **Sampling**: A token is sampled from the final probability
+      distribution using multinomial sampling.
+
+  Args:
+    logits: The unnormalized log probabilities of the vocabulary tokens,
+      with shape `[..., vocab_size]`.
```

Comment:
Nit, let's specify `...` to be `batch, sequence`? Similar comment in the return.